[grid] Eliminate vendor-specific handling of extension capabilities

sbabcoc opened the pull request:
### **User description**
### Description
The existing handling of extension capabilities assigns special significance to four recognized prefixes: `goog:`, `moz:`, `ms:`, and `se:`. Capabilities with these prefixes are entirely ignored by the current default slot matcher implementation, but custom prefixes are considered, as well as those for Safari and Appium. This inconsistency means that properties in extension capabilities can cause affected `newSession` requests to fail even though extension capabilities are supposed to be vendor-specific and should probably not be evaluated by the default slot matcher.

This PR retains the "special" status of the "se:" prefix and specifically filters out **all** extension capabilities with names that end with "options", "Options", "loggingPrefs", and "debuggerAddress". This maintains existing behavior while allowing node configurations and `newSession` requests to include extension capabilities that won't affect slot matching.

### Motivation and Context
Revolves #14461 

> Note that this is technically a breaking change, because it revises the default slot matcher to consider "special" extension capabilities without name suffix "options" that would previously have been ignored, and the matcher will now ignore "non-special" extension capabilities with name suffix "options" that would previously have been considered.
However, I'm unaware of any current use cases that will be adversely affected by this change.

The strategy employed by the PR is that extension capabilities which should be ignored for matching can be expressed as capabilities with name prefix "se:" or one of the "special" name suffixes. All other extension capabilities will be considered for slot matching. This is a generalization/extrapolation of existing patterns from current use cases.

### Recommended slot matcher enhancements

To reduce the need for custom slot matchers, we could extend the **WebDriverInfo** interface to add a new method that vendors could implement to evaluate their own criteria:

```
Boolean matches(Capabilities stereotype, Capabilities capabilities);
```

The default implementation would return `null` to indicate that no evaluation has been performed. If the vendor chooses to implement the `matches` method, their initial step must be to invoke their own `isSupporting` method, returning `null` if the corresponding driver doesn't support the specified capabilities. 

The implementation in **DefaultSlotMatcher** would be updated to iterate over the available **WebDriverInfo** providers, retaining only those whose `isSupporting` methods return `true`. The `matches` methods of this filtered list of providers will then be applied to the available nodes to determine which of these satisfies the criteria specified by the client request. The nodes that satisfy these evaluations (if any) would then be evaluated against the remaining common criteria.

Another potential enhancement would be to enable vendor-specific `matches` methods to return a weighted integer result instead of a simple `Boolean`. This would enable best-match behavior. Perhaps this is getting too complicated, though.

### Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
- [X] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.


___

### **PR Type**
Bug fix, Tests


___

### **Description**
- Removed the special status of certain extension capability prefixes in `DefaultSlotMatcher`, opting to ignore all extension capabilities with names ending in "options" (case-insensitive).
- Simplified the logic for matching capabilities by removing unnecessary checks and conditions.
- Updated `RelaySessionFactory` to streamline session creation by removing redundant capability filtering.
- Revised test cases in `DefaultSlotMatcherTest` to align with the new handling of extension capabilities.



___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>DefaultSlotMatcher.java</strong><dd><code>Revise extension capabilities handling in DefaultSlotMatcher</code></dd></summary>
<hr>

java/src/org/openqa/selenium/grid/data/DefaultSlotMatcher.java

<li>Removed special handling for specific extension capability prefixes.<br> <li> Updated logic to ignore all extension capabilities with names ending in "options" (case-insensitive). <li> Simplified capability matching logic.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14485/files#diff-08612f5a21ea03efdd38ead8885dd4b72be486ea76f0de8f0a4dce609f34af0d">+31/-51</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>RelaySessionFactory.java</strong><dd><code>Simplify session creation in RelaySessionFactory</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

java/src/org/openqa/selenium/grid/node/relay/RelaySessionFactory.java

<li>Removed logic to filter out browserName when appium:app is present.<br> <li> Simplified session creation process.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14485/files#diff-8fe345d7ed627b58287d8a4c5d6813fae678894be658c5c87ffc3f9c862f62ab">+0/-11</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>DefaultSlotMatcherTest.java</strong><dd><code>Update DefaultSlotMatcher tests for revised capability handling</code></dd></summary>
<hr>

java/test/org/openqa/selenium/grid/data/DefaultSlotMatcherTest.java

<li>Updated test cases to reflect changes in extension capability <br>handling.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14485/files#diff-efbc947cb2db3c8fe94ef12f0e57d80901e86d3e3992611966b165c1e9543445">+10/-8</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

sbabcoc commented on java/src/org/openqa/selenium/grid/data/DefaultSlotMatcher.java:
```diff
@@ -93,25 +84,21 @@ public boolean matches(Capabilities stereotype, Capabilities capabilities) {
-        .filter(name -> !"platformName".equalsIgnoreCase(name))
-        .map(
-            name -> {
-              if (capabilities.getCapability(name) instanceof String) {
```
The implementation of this conditional fails to consider that the corresponding stereotype capability could have a non-String value (including `null`), which would produce unexpected behavior. The revised implementation will only perform the case-insensitive comparison if the capability in the session request and the node stereotype both have String values.

sbabcoc commented on java/src/org/openqa/selenium/grid/data/DefaultSlotMatcher.java:
```diff
@@ -145,29 +132,22 @@ private Boolean platformVersionMatch(Capabilities stereotype, Capabilities capab
-        .orElse(true);
+      .filter(name -> name.contains(":"))
+      .filter(name -> capabilities.getCapability(name) != null)
+      .map(name -> {
```
This revised comparison will only evaluate values that are strings, numbers, or Booleans; complex values (objects and arrays) are ignored. The prior "special" treatment of four recognized extension capability prefixes (`goog:`, `moz:`, `ms:`, and `se:`) has been replaced by this consistent handling of all extension capabilities.

sbabcoc commented on java/src/org/openqa/selenium/grid/node/relay/RelaySessionFactory.java:
```diff
@@ -149,15 +147,6 @@ public Either<WebDriverException, ActiveSession> apply(CreateSessionRequest sess
               "New session request capabilities do not " + "match the stereotype."));
     }
 
-    // remove browserName capability if 'appium:app' is present as it breaks appium tests when app
```
The `merge` operation that necessitated this special-case handling has been eliminated. The session factory shouldn't be altering clients' requested capabilities.

diemol replied:
Where did it get eliminated?

sbabcoc replied:
Deleted line 160 below.

sbabcoc commented on java/test/org/openqa/selenium/grid/data/DefaultSlotMatcherTest.java:
```diff
@@ -433,10 +436,10 @@ void vendorExtensionPrefixedCapabilitiesAreIgnoredForMatching() {
-            "amsterdam",
-            "ms:fruit",
-            "mango");
+            "food:dairy",
```
This test has been updated to verify the revised behavior, in which all extension capabilities with complex values are ignored for purposes of node matching.

pujagani replied:
I think it is a good idea to create a new test and let the existing test be as it is to verify the changes don't cause a breaking change for the users.

sbabcoc replied:
I restored the original test, altering the method name and inverting the expectation to correspond with the revised behavior.

sbabcoc commented on java/test/org/openqa/selenium/grid/node/config/NodeOptionsTest.java:
```diff
@@ -148,7 +149,7 @@ boolean isDownloadEnabled(WebDriverInfo driver, String customMsg) {
               return Collections.singleton(HelperFactory.create(config, caps));
             });
-    String expected = driver.getDisplayName();
+    String expected = "Edge".equals(driver.getDisplayName()) ? Browser.EDGE.browserName() : driver.getDisplayName();
```
This test was broken when I got here. For **Microsoft Edge**, the display name is `Edge`, but the browser name is `MicrosoftEdge`. Consequently, the expected value would never be matched.

pujagani commented:
@diemol @krmahadevan Please help review this when time permits!

joerg1985 commented:
@VietND96 sorry i don't know enought to really review the changes in this PR.

From the comment in the code [here](https://github.com/SeleniumHQ/selenium/blob/751bacb6bc934436ec9dec2416a022d8d577e30a/java/src/org/openqa/selenium/grid/data/DefaultSlotMatcher.java#L148-L152) and after looking into the issue #14461 i am asking myself: should the request really contain these capabilities?

The node should be picked by the desired capability, but is able to respond with a extended set of capabilities.
To stick with the example, i request a "Appium" session and the response is this is a "Appium session started via PM2".
So the client could handle the "PM2" sessions different, but i think a intermediary node is the better place to handle this.
I don't think there is currently a extension point (like specifiying your own slot matcher) to handle the shutdown of a session.

@sbabcoc Is your custom slot matcher starting a new process and this must be terminated?

sbabcoc commented:
@joerg1985 As I read the W3C specification, the slot matcher should not be interpreting custom capabilities in any way. These should be entirely ignored and passed on without examination. Unfortunately, current functionality violates this specification. This PR modifies the slot matching behavior, replacing special-case handling of four specific prefixes with indifference to custom capabilities with complex values. This allows Appium custom capabilities to be considered for slot matching, while custom capabilities for Chrome, Edge, and Firefox will be ignored.

joerg1985 commented on java/src/org/openqa/selenium/grid/data/DefaultSlotMatcher.java:
```diff
@@ -145,27 +135,22 @@ private Boolean platformVersionMatch(Capabilities stereotype, Capabilities capab
+              return true;
             })
         .reduce(Boolean::logicalAnd)
         .orElse(true);
```
I would change this to make it more clear what is going on, filtering the element to exclude them from matching.
E.g. in pseudo code:
```
return stereotype.getCapabilityNames().stream()
   .filter(name -> name.contains(":"))
   // match only when the capabilities of the new session request contains that specific extension capability
   .filter(name -> capabilities.getCapability(name) != null)
   // {comment on why to exclude them from matching} 
   .filter(name -> !(capabilities.getCapability(name) instanceof Map))
   .filter(name -> !(capabilities.getCapability(name) instanceof List))
   .allMatch( … {same code as above in the .map of the DefaultSlotMatcher.matches method is used}
```

The `DefaultSlotMatcher.matches` method could be reworked to use `.allMatch` too, this makes it - at least for me - easier to read than `.reduce(Boolean::logicalAnd).orElse(true);`

sbabcoc replied:
@joerg1985 Thanks for the feedback! I switched all three instances of `map`/`reduce` to `allMatch` and added a few code comments.

diemol commented:
> @joerg1985 As I read the W3C specification, the slot matcher should not be interpreting custom capabilities in any way. These should be entirely ignored and passed on without examination. Unfortunately, current functionality violates this specification. This PR modifies the slot matching behavior, replacing special-case handling of four specific prefixes with indifference to custom capabilities with complex values. This allows Appium custom capabilities to be considered for slot matching, while custom capabilities for Chrome, Edge, and Firefox will be ignored.

@sbabcoc This is only valid for end nodes (ChromeDriver, GeckoDriver, etc.), Grid, and all its components are an intermediate node. That part of the spec is relaxed so the code can be customized as needed, and with that, Grid and cloud vendors can adapt to the needs of their users.

sbabcoc commented:
@diemol The main issues with the existing implementation are that it treats the "non-special" extension capabilities different than it does the "special" ones, and it alters the requested client capabilities before forwarding them to the node. My revised implementation eliminates both of these issues while retaining all currently utilized behavior.

VietND96 commented:
As my understanding, after this change, for example a Node set stereotypes `"se:containerName": "my-node1"`. In client code, it must set capability `se:containerName` to request for the slot matching, since it is a String cap, right?

sbabcoc commented:
@VietND96 Capabilities with "simple" values (Boolean/Number/String) will be considered for slot matching, even if their capability names specify a prefix. Capabilities with "complex" values (Array/Object) will be ignored for slot matching if their capability names specify a prefix.
We shouldn't give special treatment to any prefixes, because this produces inconsistent behavior. We also shouldn't be altering clients' requested capabilities in the node matching process.

VietND96 commented:
I added few tests to DefaultSlotMatcher for some scenarios that I want to verify against this change. Let's wait for the CI.

VietND96 reviewed (changes requested):
It looks like there are tests get impacted by this change. Need to fix CI.

VietND96 commented:
CI failed due to a case, where Node stereotype `moz:debuggerAddress: 127.0.0.1:3922`, and request capabilities `moz:debuggerAddress: true`
As you mentioned this change will consider caps with simplex value in slot matcher. And session could not be assigned to slot due to this change removing filter list of special prefixed extension caps `"goog:", "moz:", "ms:", "se:"` out.
As I understand, those prefixed caps will contain configs of browser, or Selenium function. In Node stereotypes it could be added automatically or assigned another value based on what it fetches from ENV, OS, etc.
And the same prefixed caps in client request caps, it could be just a boolean value to enable feature.

VietND96 commented:
I will collect and convert those failure to unit test, you just need to fix and make UT can pass when running
> bazel test //java/test/org/openqa/selenium/grid/data/...

diemol commented:
The only change we need is to avoid matching prefixed capabilities with `options` as a name. 

As in, we would not process `nord:options`, simply pass it along. 

Can you simplify the changes made in this PR to implement that, please?

sbabcoc commented:
@VietND96 I'm trying to understand why the comparison in the test you added should return `true`. Not only are the values for `moz:debuggerAddress` not the same; the values are of completely different types. I understand that this comparison would return `true` in the current implementation, but this is actually one of the bugs my PR is addressing.
I strongly suspect that there are no real-world scenarios in which this sort of mismatch should be ignored. This seems like a test that's been tailored to verify an arbitrary behavior that's an unintended artifact of the current implementation. If the extension prefixes were replaced with `appium`, the current implementation returns `false`, and this is a behavior that existing clients actually depend on.

sbabcoc commented:
@diemol Changing the behavior to only ignore extension capabilities whose names contain "options" would produce the same behavior that is causing the test @VietND96 added to fail. The values of `moz:debuggerAddress` would be compared and found to be unequal, and the comparison would return `false`. As I stated in my previous reply, I believe this is the correct behavior - unequal "simple" values should cause the comparison to return `false`.

The approach I'm presenting in this PR provides the greatest level of control and consistency for clients.
* If the client wants the capability to be considered for matching, declare it with a "simple" value.
* If the client wants the capability to be ignored for matching, declare it with a "complex" value.
* No extension capability prefixes get "special" treatment, which avoids confusing, inconsistent behavior for custom extensions.

diemol commented:
>The values of moz:debuggerAddress would be compared and found to be unequal, and the comparison would return false.

No, because we will still ignore `moz:`. The only change we need is to ignore `:options`, which makes sense to be ignored when matching.

diemol commented:
All other changes should be reverted, please.

sbabcoc commented:
@diemol I understand that we currently ignore all "special" extension capabilities, but there appears to be no real-world scenarios of "special" extension capabilities with "simple" values in which this behavior is actually needed. For Appium extension capabilities, current clients rely on capabilities with "simple" values being considered for node matching.
I believe that the existing treatment of "special" extension capabilities is undesirably permissive, and the inconsistent behavior we see for "non-special" extension capabilities is also undesirable.

diemol commented:
>I believe that the existing treatment of "special" extension capabilities is undesirably permissive, and the inconsistent behavior we see for "non-special" extension capabilities is also undesirable.

Why?

sbabcoc commented:
@diemol Why is it desirable to support the scenario enforced by the test that @VietND96 added? We match nodes even when capabilities like `moz:debuggerAddress` are entirely different between requested and actual values. We don't exhibit this behavior for Appium nodes, where capabilities like `appium:automationName` are considered when matching requested and actual values.
The revised behavior you recommend (only ignoring extension capabilities whose names contain "options") would produce behavior that doesn't satisfy the condition expected by the test that @VietND96 added.

diemol commented:
I cannot see what @VietND96 added because you are force-pushing the whole time, and I don't know who added what. I don't know which scenario you mean.

VietND96 commented:
I added it back to trunk https://github.com/SeleniumHQ/selenium/commit/10119a9094e290c3720f9b4f8dece0919a2cd850

diemol commented:
@sbabcoc is correct, this `moz:debuggerAddress` should not be considered for matching. That is an internal Firefox thing.

sbabcoc commented:
@diemol I force-push after rebasing to keep the commit history clean. My attempt to rebase after @VietND96 added his unit test went horribly wrong and I had to check out a prior commit to recover, which removed his test. I let him know, and he created a separate PR to add his test to trunk.

sbabcoc commented:
@diemol It seems like you're advocating for ignoring all extension capabilities for purposes of slot matching. This approach would mean that custom matchers would be required in all scenarios where matching based on extension capabilities is desired (e.g. - `appium:automationName`). The approach implemented by this PR is a compromise. It enables clients add slot matching criteria by adding extension capabilities with "simple" values, while extension capabilities specified as "complex" will be ignored.

The enhancements I recommended in the PR description provide the outline of a mechanism for vendors to define slot matcher extensions that evaluate their specific matching criteria. This would enable vendors to define their own custom criteria without the need to implement custom slot matchers.

diemol commented:
>@diemol It seems like you're advocating for ignoring all extension capabilities for purposes of slot matching.

Not really. Just the ones named `options`, as in `nord:options`.

sbabcoc commented:
@diemol I can revise my PR to implement this strategy. It is significantly less complex to implement and much easier to describe. I'll update the unit tests accordingly.

Is the concept of a slot matcher extension mechanism worth exploring further? This would be an additional method for the **WebDriverInfo** interface, which is already being scanned for service providers.

diemol commented:
You can already implement your own [SlotMatcher](https://www.selenium.dev/documentation/grid/configuration/cli_options/#distributor), see `--slot-matcher`

sbabcoc commented:
@diemol Yes, but you can only specify one custom matcher. If you need sessions from multiple vendors, this means running with multiple hubs. With a slot matcher extension mechanism, vendor-specific matching gets activated automatically and standing up a grid that supplies sessions from multiple vendors is much easier.

diemol commented:
The matcher is used both at the Distributor and the Node. You would need to decouple that logic to allow more matchers.

sbabcoc commented:
@diemol The extension would be implemented in **DefaultSlotMatcher**, which is already integrated with Distributor and Node.

sbabcoc commented:
@diemol I've revised the matcher to only ignore extension capabilities with suffix "options" (case-insensitive) and updated the affected unit test accordingly. I also updated the PR description to indicate the revised implementation.

diemol commented on java/src/org/openqa/selenium/grid/data/DefaultSlotMatcher.java:
```diff
@@ -44,13 +42,6 @@
-   List of prefixed extension capabilities we never should try to match, they should be
-   matched in the Node or in the browser driver.
-  */
-  private static final List<String> EXTENSION_CAPABILITIES_PREFIXES =
```
I want to keep this because the browser vendors can make changes and we don't want to change things here every time they do it. For example, `moz:debuggerAddress` is now being sent from the client side to enable CDP in Firefox.

sbabcoc replied:
The core objective of this PR is to migrate all vendor-specific capabilities that should be ignored for purposes of slot matching into "options" objects. (I had initially implemented this to ignore all extension capabilities with "complex" values.) The current strategy of ignoring all extension capabilities with a small number of "special" prefixes while considering all extension capabilities with "non-special" prefixes produces inconsistent behavior and requires vendors to implement special-case slot matchers, each of which is mutually exclusive to every other custom slot matcher.

In your example of `moz:debuggerAddress`, the desire is to ignore this capability for slot matching. What happens if Mozilla adds a new extension capability that should be considered for slot matching? The current strategy doesn't allow for this possibility, forcing Mozilla to implement a custom slot matcher, and for all affected clients to update their Grid configurations to add this matcher. For clients with Grid configurations that supply sessions of browsers from multiple vendors, this could force complete reconfiguration of their automation infrastructure, segregating their session providers into multiple grids by vendor.

With the revised behavior implemented by this PR, `moz:debuggerAddress` would be considered for slot matching. To ignore this option for slot matching, it would be migrated into `moz:firefoxOptions`, I believe this is a more sensible approach, because this capability truly is an option, not an identity specifier.

diemol replied:
>The current strategy of ignoring all extension capabilities with a small number of "special" prefixes while considering all extension capabilities with "non-special" prefixes produces inconsistent behavior and requires vendors to implement special-case slot matchers, each of which is mutually exclusive to every other custom slot matcher

This is an assumption.


If Mozilla adds a new capability for matching, we ignore it and pass it along to GeckoDriver.

sbabcoc replied:
This is precisely my point. Outside of mutually exclusive custom slot matchers, vendors have no control over how their extension capabilities are handled. Capabilities with "special" prefixes will always be ignored and capabilities with "non-special" prefixes will always be considered. This PR eliminates the inconsistency and adds the ability for vendors to control whether their extension capabilities will be treated as identifying characteristics or as end-node configuration options.

diemol replied:
Please leave the prefixes. Then we can merge the PR.

sbabcoc replied:
I revised my implementation to restore the "special" status of extension capabilities with prefix "se:". I also added explicit filtering for extension capabilities with suffixes "options", "Options", "loggingPrefs", and "debuggerAddress". This captures all existing use cases, which avoids breaking any existing functionality.
I had failed to identify the full collection of existing extension capabilities that needed to be ignored for slot matching. Thanks for your reviews and feedback!

diemol replied:
Please leave the prefixes. Then we can merge the PR.

diemol replied:
@sbabcoc I won't merge the PR because the comments have not been addressed. The request is to leave the prefixes as they were and only focus on ignoring capabilities with a prefix and the name ` options`.

sbabcoc replied:
@diemol I understand your request to reinstate the "special" handling of all extension capabilities with prefixes `goog`, `moz`, and `ms`. Doing so would reintroduce the inconsistency that I'm attempting to eliminate.

In its current form, the revisions in this PR eliminate the special status of `goog`, `moz`, and `ms`.  The only prefix that is entirely ignored is `se`. To continue ignoring the small number of existing vendor-specific extension capabilities, I added explicit checks for these suffixes: "options", "Options", "loggingPrefs", and "debuggerAddress".

There are two capability categories: identity and configuration. The former is considered for slot matching; the latter is ignored for slot matching. The existing implementation of **DefaultSlotMatcher** handles extension capabilities inconsistently, ignoring all capabilities from three "special" vendors (configuration) while considering all other capabilities (identity). Vendors (even the "special" ones) lack the ability to differentiate identity values from configuration values.

If my PR is approved and merged, vendors will have the ability to add identity values and configuration values as needed, with the guarantee that the default slot matcher will treat them as intended.

diemol replied:
Why is it an inconsistency? Browser vendors just need those capabilities to be passed along. They do not use the Grid and configure stereotypes to do matching.

sbabcoc replied:
Every extension capability for Google, Mozilla, and Microsoft is ignored for slot matching; every extension capability for everyone else (including Apple and Appium) is considered for slot matching. This behavior is inconsistent. It also fails to provide vendors with the ability to control which capabilities are treated as identity values and which are treated as configuration options.

diemol replied:
Apple does not need to be there; this might change in the future.

However, Appium does need to be considered, except the `options` one.

Browser vendors just need to have their capabilities passed along. Why do you keep saying that they will use their capabilities to do matching in Grid when this is not true?

sbabcoc replied:
Appium extension capabilities are most certainly being considered for slot matching. This is still true in the revised functionality implemented by this PR. I think the behavior of the current matcher is what necessitated the merge operation and explicit check for 'appium:app'.

sbabcoc replied:
With my current implementation, all existing Appium extension capabilities are considered except for the four suffixes: "options", "Options", "loggingPrefs", and "debuggerAddress". There's implementation in the Appium server that merges the prefixed capabilities into their "options" collection with no thought to which capabilities are "identity" values and which are configuration options. I don't know if this gets reflected back to Grid in a way that's visible to **DefaultSlotMatcher**, but their configuration examples indicate that every option can be declared in either place. I'm attempting to engage the Appium folks to see how they feel about deprecating this pattern in favor of clear differentiation between "identity" values and configuration options.

diemol replied:
I would not prefer to leave `loggingPrefs`, and `debuggerAddress` hardcoded. I prefer to have the 4 prefixes ignored as it was before.

I don't mean to be rude, I am taking your considerations, and I strongly prefer the approach I've been suggesting in previous comments. Please let me know if that works for you or if I should raise the change in a different PR.

sbabcoc replied:
Thanks for remaining engaged in this conversation! I understand your hesitation to alter the established behavior. Perhaps the new behavior can be activated by an optional core Selenium capability. That way, the new behavior will only be present when it's explicitly requested.

diemol commented on java/src/org/openqa/selenium/grid/node/relay/RelaySessionFactory.java:
```diff
@@ -149,15 +147,6 @@ public Either<WebDriverException, ActiveSession> apply(CreateSessionRequest sess
-      filteredStereotype.setCapability(CapabilityType.BROWSER_NAME, (String) null);
-    }
-
-    capabilities = capabilities.merge(filteredStereotype);
```
Now I remember. This was done because the user can specify something in the stereotype that will be sent to the Node that will receive the request.
